### PR TITLE
implement ImmutableQueue and basic unit tests

### PR DESCRIPTION
**Pull request by rui-gao**

### CODE REVIEW

File: src/test/java/ImmutableQueueTest.java [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+        assertEquals(0, immutableQueueWithOneElement.head().getValue());
+
+        // enQueue the second element
+        ImmutableQueue<MyImmutableElement> immutableQueueWithTwoElement =
+                (ImmutableQueue<MyImmutableElement>) immutableQueueWithOneElement.enQueue(new MyImmutableElement(1));
+        // assert throw RuntimeException(QueueFullException) and exception message when try to enQueue the third element
+        QueueFullException expectedQueueFullException =
+                assertThrows(QueueFullException.class,
+                        () -> immutableQueueWithTwoElement.enQueue(new MyImmutableElement(2)));
+        assertEquals(
```
Comment (rui-gao) [inline src/test/java/ImmutableQueueTest.java:40]: Edge case covered

File: src/test/java/ImmutableQueueTest.java [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+        // assert the second element in original queue becomes the first element in the new queue
+        ImmutableQueue<MyImmutableElement> immutableQueueDeQueueOneElement =
+                (ImmutableQueue<MyImmutableElement>) immutableQueueWithTwoElement.deQueue();
+        assertEquals(1, immutableQueueDeQueueOneElement.head().getValue());
+
+        // DeQueue the first element
+        // assert the new queue is empty
+        ImmutableQueue<MyImmutableElement> immutableQueueDeQueueTwoElement =
+                (ImmutableQueue<MyImmutableElement>) immutableQueueDeQueueOneElement.deQueue();
+        assertTrue(immutableQueueDeQueueTwoElement.isEmpty());
```
Comment (rui-gao) [inline src/test/java/ImmutableQueueTest.java:64]: Edge case covered

File: src/test/java/ImmutableQueueTest.java [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+    }
+
+    @Test
+    public void head() {
+        // Make a empty queue
+        // assert head() return null
+        ImmutableQueue<MyImmutableElement> immutableQueue = new ImmutableQueue<>(10);
+        assertNull(immutableQueue.head());
```
Comment (rui-gao) [inline src/test/java/ImmutableQueueTest.java:77]: Edge case covered

File: src/test/java/ImmutableQueueTest.java [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+        // assert head() return the right element in a no empty queue
+        ImmutableQueue<MyImmutableElement> immutableQueueWithOneElement =
+                (ImmutableQueue<MyImmutableElement>) immutableQueue.enQueue(new MyImmutableElement(0));
+        ImmutableQueue<MyImmutableElement> immutableQueueWithTwoElement =
+                (ImmutableQueue<MyImmutableElement>) immutableQueueWithOneElement.enQueue(new MyImmutableElement(1));
+        assertEquals(0, immutableQueueWithTwoElement.head().getValue());
+    }
+
+    @Test
+    public void isEmpty() {
```
Comment (rui-gao) [inline src/test/java/ImmutableQueueTest.java:88]: branch cases covered

File: src/test/java/ImmutableQueueTest.java [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+        assertTrue(immutableQueue.isEmpty());
+
+        // assert isEmpty() return false for no empty queue
+        ImmutableQueue<MyImmutableElement> immutableQueueWithOneElement =
+                (ImmutableQueue<MyImmutableElement>) immutableQueue.enQueue(new MyImmutableElement(0));
+        ImmutableQueue<MyImmutableElement> immutableQueueWithTwoElement =
+                (ImmutableQueue<MyImmutableElement>) immutableQueueWithOneElement.enQueue(new MyImmutableElement(1));
+        assertFalse(immutableQueueWithTwoElement.isEmpty());
+    }
+}
```
Comment (rui-gao) [inline src/test/java/ImmutableQueueTest.java]: TODO: need new line

File: src/test/java/ImmutableQueueTest.java [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+        QueueFullException expectedQueueFullException =
+                assertThrows(QueueFullException.class,
+                        () -> immutableQueueWithTwoElement.enQueue(new MyImmutableElement(2)));
+        assertEquals(
+                "ImmutableQueue capacity exceed, can NOT add more element, capacity is 2",
+                expectedQueueFullException.getMessage());
+    }
+
+    @Test
+    public void deQueue() {
```
Comment (rui-gao) [inline src/test/java/ImmutableQueueTest.java:46]: TODO: add deQueue empty queue case.

**Comment by rui-gao** [comment]
TODO: should make the ImmutableQueue Thread-Safety

TODO: should optimize the performance of the ImmutableQueue

**Comment by rui-gao** [comment]
> TODO: should optimize the performance of the ImmutableQueue
TODO: should optimize the performance of the ImmutableQueue

http://www.javacreed.com/comparing-the-performance-of-various-list-implementations/ 
Seems ArrayList is a good choice for the container if not consider the potential operation for expending ArrayList inside array when it's size get big. Leaving advanced optimization as future work then.

>TODO: should make the ImmutableQueue Thread-Safety

Added synchronized to the methods of the ImmutableQueue class. But, now I realized that we don't need to synchronize the methods actually. Because we make the queue immutable, once created the capacity and the list won't be modified at all. So, will delete synchronized for good.
